### PR TITLE
Add `MetricInputTransformer`

### LINKED ISSUE #2371: MetricWrapper for Target Binarization
## 🚀 Feature

Add a `TargetBinarizationWrapper` that cast continuous labels to binary labels given a threshold.

### Motivation

Evaluating two metrics that require different label formats (e.g., one binary, the other continuous) is cumbersome since it requires setting up two different evaluation stacks where one is fed with binarized label data and the other is fed the original continuous data. This leads to code duplication. Also, persisting binarized labels into the dataset in scenarios where a metric requires different input than what is given in the ground-truth data diminishes code clarity w.r.t. the evaluation process.

### Pitch

A metric wrapper that casts target data to binary targets during the `.update()` method. Can be applied to either a single `Metric`, or a whole `MetricCollection`.

### Alternatives

- using a `MultiTaskWrapper` is possible, but has two caveats: (1) metrics with a different signature than `update(pred, target)` are not supported, and (2) it requires the user to implement the thresholding logic by themselves before feeding it into the `MultiTaskWrapper`
- use a more generic target processing wrapper that would allow supplying, e.g., a custom `lambda` that is applied to targets; more flexible, but also more complex to implement

### PR DESCRIPTION
**Pull request by lgienapp**
## What does this PR do?

Fixes #2371 

This PR adds a new kind of metric wrapper for data transformations. Transformations allow for modifications to the input a metric receives by wrapping its `pred` and `target` arguments. Transformations can be implemented by either subclassing the `MetricInputTransformer` base class and overriding the `.transform_pred()` and/or `.transform_target()` functions, or by supplying a lambda function via the `LambdaInputTransformer`. A `BinaryTargetTransformer` which casts target labels to 0/1 given a threshold is provided for convenience.

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

----
📚 Documentation preview 📚: https://torchmetrics--2392.org.readthedocs.build/en/2392/

### CODE REVIEW

**Comment by lgienapp** [comment]
Unsure what to do about the failing [mypy test]() - assigning the lambdas to overwrite the original class methods seems to be the most straightforward solution to supplying user-defined behaviour here; I would add the file to `[[tool.mypy.overrides]]` to mute the error (since its the only one for `transformations.py`) unless someone sees a better in-code solution?

File: docs/source/wrappers/transformations.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+.. include:: ../links.rst
+
+###############
+Transformations
+###############
+
+Transformations allow for modifications to the input a metric receives by wrapping its `pred` and `target` arguments.
+Transformations can be implemented by either subclassing the `MetricInputTransformer` base class and overriding the `.transform_pred()` and/or `transform_target()` functions, or by supplying a lambda function via the `LambdaInputTransformer`.
+A `BinaryTargetTransformer` which casts target labels to 0/1 given a threshold is provided for convenience.
```
Comment (Borda) [inline docs/source/wrappers/transformations.rst:14]:
```suggestion
Transformations allow for modifications to the input a metric receives by wrapping its `pred` and `target` arguments.
Transformations can be implemented by either subclassing the ``MetricInputTransformer`` base class and overriding the ``.transform_pred()`` and/or ``transform_target()`` functions, or by supplying a lambda function via the ``LambdaInputTransformer``.
A ``BinaryTargetTransformer`` which casts target labels to 0/1 given a threshold is provided for convenience.
```
RST files have different code annotation the MD :)

File: tests/unittests/wrappers/test_transformations.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+from typing import ClassVar
+
+import pytest
+from torch import Tensor
+from torchmetrics.aggregation import MeanMetric
+from torchmetrics.classification import BinaryAccuracy
+from torchmetrics.retrieval import RetrievalMAP
+from torchmetrics.wrappers import BinaryTargetTransformer, LambdaInputTransformer, MetricInputTransformer
+
+from unittests.helpers import seed_all
```
Comment (baskrahmer) [inline tests/unittests/wrappers/test_transformations.py:23]:
```suggestion
from unittests._helpers import seed_all
```

File: tests/unittests/wrappers/test_transformations.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        self, cls, transform_pred, transform_target, preds, preds_transformed, targets, targets_transformed
+    ) -> None:
+        """Tests if the binarized forward matches the output of the metric on manually binarized targets."""
+        metric = cls()
+        wrapped_metric = LambdaInputTransformer(
+            metric, transform_pred=transform_pred, transform_target=transform_target
+        )
+        preds = Tensor(preds).float()
+        targets = Tensor(targets).float()
+        preds_transformed = Tensor(preds_transformed).float() if preds_transformed is not None else preds
```
Comment (baskrahmer) [inline tests/unittests/wrappers/test_transformations.py:103]: `preds_transformed` and `targets_transformed` are never None, so these statements can be removed:
```suggestion
        preds_transformed = Tensor(preds_transformed).float()
```

**Review by baskrahmer — APPROVED** [review]
Looks good :)

**Review by baskrahmer — APPROVED** [review]

**Comment by Borda** [comment]
@lgienapp I just fixed some typing, but suddenly, all tests are failing, Do you mind checking what is wrong... :flamingo:

**Comment by lgienapp** [comment]
Some unit test check for error messages via regexes, and these had to be adapted to the spelling fixes as well. Should all pass now.

File: src/torchmetrics/wrappers/transformations.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    Raises:
+        TypeError:
+            If `transform_pred` is not a Callable.
+        TypeError:
+            If `transform_target` is not a Callable.
+
+    Example:
+        >>> import torch
+        >>> from torchmetrics.wrappers import LambdaInputTransformer
+        >>> from torchmetrics.classification import BinaryAccuracy
```
Comment (stancld) [inline src/torchmetrics/wrappers/transformations.py:99]: Complete nitpick, I just see `isort/ruff` everywhere :D 

```suggestion
        >>> from torchmetrics.classification import BinaryAccuracy
        >>> from torchmetrics.wrappers import LambdaInputTransformer
```

File: src/torchmetrics/wrappers/transformations.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            The binarization threshold for the targets. Targets values `t` are cast to binary with `t > threshold`.
+
+    Raises:
+        TypeError:
+            If `threshold` is not an `int` or `float`.
+
+    Example:
+        >>> import torch
+        >>> from torchmetrics.wrappers import BinaryTargetTransformer
+        >>> from torchmetrics.retrieval import RetrievalMRR
```
Comment (stancld) [inline src/torchmetrics/wrappers/transformations.py:150]: ditto

```suggestion
        >>> from torchmetrics.retrieval import RetrievalMRR
        >>> from torchmetrics.wrappers import BinaryTargetTransformer
```

File: src/torchmetrics/wrappers/transformations.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        >>> topics = torch.tensor([0,0,0,0,0,1,1,1,1,1])
+        >>>
+        >>> metric = BinaryTargetTransformer(RetrievalMRR())
+        >>> metric.update(preds, targets, indexes=topics)
+        >>> metric.compute()
+        tensor(0.7500)
+
+    """
+
+    def __init__(self, wrapped_metric: Union[Metric, MetricCollection], threshold: float = 0, **kwargs: Any) -> None:
```
Comment (stancld) [inline src/torchmetrics/wrappers/transformations.py:163]:
```suggestion
    def __init__(self, wrapped_metric: Union[Metric, MetricCollection], threshold: Union[int, float] = 0, **kwargs: Any) -> None:
```

**Review by stancld — APPROVED** [review]
Nice work, thanks for the addition!
